File: Example/XeeSDK/XeeViewController.m
//
//  XeeViewController.m
//  XeeSDK
//
//  Created by jbdujardin on 05/02/2017.
//  Copyright (c) 2017 jbdujardin. All rights reserved.
//

#import "XeeViewController.h"
#import "XeeSDK.h"

@interface Cell : UITableViewCell
    
@property (nonatomic, strong) IBOutlet UILabel *label;
    
@end

@implementation Cell
    
@end

@interface XeeViewController () <UITableViewDelegate, UITableViewDataSource, XeeConnectManagerDelegate>

@property (nonatomic, strong) NSNumber *carId;
@property (nonatomic, strong) NSString *tripId;
@property (nonatomic, strong) NSString *deviceId;
    
@property (nonatomic, strong) IBOutlet UIButton *btnCarId;
@property (nonatomic, strong) IBOutlet UIButton *btnTripId;
@property (nonatomic, strong) IBOutlet UIButton *btnDeviceId;
    
@property (nonatomic, strong) IBOutlet UITableView *tableView;

@end

@implementation XeeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    self.textView.text = @"";
    
    self.carId = @0;
    self.tripId = @"";
    self.deviceId = @"";
    
    [self.btnCarId setTitle:@"edit" forState:UIControlStateNormal];
    [self.btnTripId setTitle:@"edit" forState:UIControlStateNormal];
    [self.btnDeviceId setTitle:@"edit" forState:UIControlStateNormal];
    
    // TEST - create programmatically xee login button
    /*XeeLoginButton *button = [[XeeLoginButton alloc] init];
     button.delegate = self;
     button.style = XeeLoginButtonStyleDark;
     button.center = self.view.center;
     [self.view addSubview:button];*/
}
    
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}
    
-(void)show:(NSString*)message {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.textView.contentOffset = CGPointZero;
        self.textView.text = message;
    });
}
    
-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    Cell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    
    switch (indexPath.row) {
        case 0:
        cell.label.text = @"Connect";
        cell.backgroundColor = [UIColor colorWithRed:0.8 green:0.8 blue:0.9 alpha:1.0];
        break;
        
        case 1:
        cell.label.text = @"Me";
        cell.backgroundColor = [UIColor colorWithRed:0.9 green:0.9 blue:0.9 alpha:1.0];
        break;
        case 2:
        cell.label.text = @"Me/Cars";
        cell.backgroundColor = [UIColor colorWithRed:0.9 green:0.9 blue:0.9 alpha:1.0];
        break;
        case 3:
        cell.label.text = @"Me/Devices";
        cell.backgroundColor = [UIColor colorWithRed:0.9 green:0.9 blue:0.9 alpha:1.0];
        break;
        
        case 4:
        cell.label.text = @"Car signals";
        cell.backgroundColor = [UIColor colorWithRed:0.8 green:0.9 blue:0.8 alpha:1.0];
        break;
        case 5:
        cell.label.text = @"Car locations";
        cell.backgroundColor = [UIColor colorWithRed:0.8 green:0.9 blue:0.8 alpha:1.0];
        break;
        case 6:
        cell.label.text = @"Car geoJSON";
        cell.backgroundColor = [UIColor colorWithRed:0.8 green:0.9 blue:0.8 alpha:1.0];
        break;
        case 7:
        cell.label.text = @"Car Status";
        cell.backgroundColor = [UIColor colorWithRed:0.8 green:0.9 blue:0.8 alpha:1.0];
        break;
        case 8:
        cell.label.text = @"Car trips";
        cell.backgroundColor = [UIColor colorWithRed:0.8 green:0.9 blue:0.8 alpha:1.0];
        break;
        case 9:
        cell.label.text = @"Car used time";
        cell.backgroundColor = [UIColor colorWithRed:0.8 green:0.9 blue:0.8 alpha:1.0];
        break;
        case 10:
        cell.label.text = @"Car mileage";
        cell.backgroundColor = [UIColor colorWithRed:0.8 green:0.9 blue:0.8 alpha:1.0];
        break;
        
        case 11:
        cell.label.text = @"Trip";
        cell.backgroundColor = [UIColor colorWithRed:0.9 green:0.8 blue:0.8 alpha:1.0];
        break;
        case 12:
        cell.label.text = @"Trip location";
        cell.backgroundColor = [UIColor colorWithRed:0.9 green:0.8 blue:0.8 alpha:1.0];
        break;
        case 13:
        cell.label.text = @"Trip geoJSON";
        cell.backgroundColor = [UIColor colorWithRed:0.9 green:0.8 blue:0.8 alpha:1.0];
        break;
        case 14:
        cell.label.text = @"Trip signals";
        cell.backgroundColor = [UIColor colorWithRed:0.9 green:0.8 blue:0.8 alpha:1.0];
        break;
        
        case 15:
        cell.label.text = @"Device signals";
        cell.backgroundColor = [UIColor colorWithRed:0.4 green:0.9 blue:0.8 alpha:1.0];
        break;
        case 16:
        cell.label.text = @"Device status";
        cell.backgroundColor = [UIColor colorWithRed:0.4 green:0.9 blue:0.8 alpha:1.0];
        break;
        
        case 17:
        cell.label.text = @"disconnect";
        cell.backgroundColor = [UIColor colorWithRed:0.9 green:0.9 blue:0.9 alpha:1.0];
        break;
        
        default:
        break;
    }
    
    return cell;
}
    
-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 17;
}
    
-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    switch (indexPath.row) {
        //connect
        case 0: {
            [Xee connectManager].delegate = self;
            [[Xee connectManager] connect];
        }
        break;
        
        //me
        case 1: {
            [[Xee requestManager].users me:^(XeeUser *user, NSError *error) {
                if(!error) {
                    [self show:user.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 2: {
            [[Xee requestManager].users meCars:^(NSArray<XeeCar *> *cars, NSError *error) {
                if(!error) {
                    [self show:cars.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
            
        case 3: {
            [[Xee requestManager].users meDevices:^(NSArray<XeeDevice *> *devices, NSError *error) {
                if(!error) {
                    [self show:devices.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        //cars
        case 4: {
            [[Xee requestManager].cars signalsWithCarId:self.carId limit:@0 begin:[NSDate dateWithTimeIntervalSinceNow:-3600*10] end:nil name:nil completionHandler:^(NSArray<XeeSignal *> *signals, NSError *error) {
                if(!error) {
                    [self show:signals.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 5: {
            [[Xee requestManager].cars locationsWithCarId:self.carId limit:@0 begin:[NSDate dateWithTimeIntervalSinceNow:-3600*10] end:nil completionHandler:^(NSArray<XeeLocation *> *locations, NSError *error) {
                if(!error) {
                    [self show:locations.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 6: {
            [[Xee requestManager].cars locationsGeoJSONWithCarId:self.carId limit:@0 begin:[NSDate dateWithTimeIntervalSinceNow:-3600*10] end:nil completionHandler:^(NSArray *locations, NSError *error) {
                if(!error) {
                    [self show:locations.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 7: {
            [[Xee requestManager].cars statusWithCarId:self.carId completionHandler:^(XeeCarStatus *carStatus, NSError *error) {
                if(!error) {
                    [self show:carStatus.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 8: {
            [[Xee requestManager].cars tripsWithCarId:self.carId begin:[NSDate dateWithTimeIntervalSinceNow:-3600*24*7] end:nil completionHandler:^(NSArray<XeeTrip *> *trips, NSError *error) {
                if(!error) {
                    [self show:trips.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 9: {
            [[Xee requestManager].stats usedTimeWithCarId:self.carId begin:nil end:nil initialValue:@0 completionHandler:^(XeeStat *stat, NSError *error) {
                if(!error) {
                    [self show:stat.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 10: {
            [[Xee requestManager].stats mileageWithCarId:self.carId begin:nil end:nil initialValue:@0 completionHandler:^(XeeStat *stat, NSError *error) {
                if(!error) {
                    [self show:stat.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        //trips
        case 11: {
            [[Xee requestManager].trips tripWithId:self.tripId completionHandler:^(XeeTrip *trip, NSError *error) {
                if(!error) {
                    [self show:trip.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 12: {
            [[Xee requestManager].trips locationsWithTripId:self.tripId completionHandler:^(NSArray<XeeLocation *> *locations, NSError *error) {
                if(!error) {
                    [self show:locations.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 13: {
            [[Xee requestManager].trips locationsGeoJSONWithTripId:self.tripId completionHandler:^(NSArray *locations, NSError *error) {
                if(!error) {
                    [self show:locations.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 14: {
            [[Xee requestManager].trips signalsWithTripId:self.tripId name:nil completionHandler:^(NSArray<XeeSignal *> *signals, NSError *error) {
                if(!error) {
                    [self show:signals.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        //devices
        case 15: {
            [[Xee requestManager].device signalsWithDeviceId:self.deviceId limit:@0 begin:nil end:nil name:nil completionHandler:^(NSArray<XeeSignal *> *signals, NSError *error) {
                if(!error) {
                    [self show:signals.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 16: {
            [[Xee requestManager].device deviceStatusWithDeviceId:self.deviceId completionHandler:^(XeeDeviceStatus *deviceStatus, NSError *error) {
                if(!error) {
                    [self show:deviceStatus.description];
                } else {
                    [self show:error.description];
                }
            }];
        }
        break;
        
        case 17:
        [[Xee connectManager] disconnect];
        break;
        
        default:
        break;
    }
}
    
-(IBAction)btnSetCarIdHandler:(id)sender {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"" message:@"Set your test car id" preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.text = [NSString stringWithFormat:@"%@", self.carId];
    }];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        UITextField *textField = alert.textFields[0];
        self.carId = [NSNumber numberWithInteger:[textField.text integerValue]];
        [self.btnCarId setTitle:[NSString stringWithFormat:@"%@", self.carId] forState:UIControlStateNormal];
    }];
    [alert addAction:action];
    [self presentViewController:alert animated:YES completion:nil];
}
    
-(IBAction)btnSetTripIdHandler:(id)sender {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"" message:@"Set your test trip id" preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.text = self.tripId;
    }];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        UITextField *textField = alert.textFields[0];
        self.tripId = textField.text;
        [self.btnTripId setTitle:self.tripId forState:UIControlStateNormal];
    }];
    [alert addAction:action];
    [self presentViewController:alert animated:YES completion:nil];
}

- (IBAction)btnSetDeviceIdHandler:(id)sender {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"" message:@"Set your test device id" preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.text = self.deviceId;
    }];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        UITextField *textField = alert.textFields[0];
        self.deviceId = textField.text;
        [self.btnDeviceId setTitle:self.deviceId forState:UIControlStateNormal];
    }];
    [alert addAction:action];
    [self presentViewController:alert animated:YES completion:nil];
}
    
-(UIView *)connectManagerViewForSignUp {
    return self.view;
}
    
-(UIView *)connectManagerViewForLogin {
    return self.view;
}
    
-(void)connectManager:(XeeConnectManager *)connectManager didFailWithError:(NSError *)error {
    [self show:error.description];
}
    
-(void)connectManager:(XeeConnectManager *)connectManager didSuccess:(XeeAccessToken *)accessToken {
    [self show:accessToken.description];
}
    
-(void)connectManagerDidCancel:(XeeConnectManager *)connectManager {
    
}

@end

File: Example/XeeSDK/XeeViewController.swift
//
//  XeeViewController.swift
//  XeeSDK_Example
//
//  Created by Jean-Baptiste Dujardin on 03/10/2017.
//  Copyright © 2017 CocoaPods. All rights reserved.
//

import UIKit
import XeeSDK

class XeeViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, XeeConnectManagerDelegate {
    
    @IBOutlet var textView: UITextView!
    @IBOutlet var userIDLabel: UILabel!
    @IBOutlet var carIDLabel: UILabel!
    @IBOutlet var tripIDLabel: UILabel!
    @IBOutlet var deviceIDLabel: UILabel!
    
    var userID: String? {
        didSet {
            userIDLabel.text = "userID : \(userID!)"
        }
    }
    var vehiculeID: String? {
        didSet {
            carIDLabel.text = "carID : \(vehiculeID!)"
        }
    }
    var tripID: String? {
        didSet {
            tripIDLabel.text = "tripID : \(tripID!)"
        }
    }
    var deviceID: String? {
        didSet {
            deviceIDLabel.text = "deviceID : \(deviceID!)"
        }
    }
    var privacyID: String?

    override func viewDidLoad() {
        super.viewDidLoad()
        self.textView.text = ""
        // Do any additional setup after loading the view.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - UITableViewDataSource
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 16
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "basicCell", for: indexPath)
        
        switch indexPath.row {
        case 0:
            cell.textLabel?.text = "Create Account"
            break
        case 1:
            cell.textLabel?.text = "Connect"
            break
        case 2:
            cell.textLabel?.text = "Users/me"
            break
        case 3:
            cell.textLabel?.text = "Update User"
            break
        case 4:
            cell.textLabel?.text = "Vehicles"
            break
        case 5:
            cell.textLabel?.text = "Vehicle"
            break
        case 6:
            cell.textLabel?.text = "Vehicle Status"
            break
        case 7:
            cell.textLabel?.text = "Update Vehicle"
            break
        case 8:
            cell.textLabel?.text = "Device"
            break
        case 9:
            cell.textLabel?.text = "Get Privacies"
            break
        case 10:
            cell.textLabel?.text = "Start Privacy"
            break
        case 11:
            cell.textLabel?.text = "Stop Privacy"
            break
        case 12:
            cell.textLabel?.text = "Trips"
            break
        case 13:
            cell.textLabel?.text = "Trip"
            break
        case 14:
            cell.textLabel?.text = "Trip Signals"
            break
        case 15:
            cell.textLabel?.text = "Disconnect"
            break
        default:
            break
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        switch indexPath.row {
        // Create Account
        case 0:
            XeeConnectManager.shared.delegate = self
            XeeConnectManager.shared.createAccount()
            break
        // Connect
        case 1:
            XeeConnectManager.shared.delegate = self
            XeeConnectManager.shared.connect()
            break
        // User/me
        case 2:
            XeeRequestManager.shared.getUser(completionHandler: { (error, user) in
                if let error = error {
                    self.textView.text = error.localizedDescription
                }else if let user = user {
                    self.userID = user.userID
                    self.textView.text = user.toJSONString(prettyPrint: true)
                }
            })
            break
        // Update User
        case 3:
            XeeRequestManager.shared.getUser(completionHandler: { (error, user) in
                if let user = user {
                    let temp = user.firstName
                    user.firstName = "TEST"
                    XeeRequestManager.shared.updateUser(WithUser: user, completionHandler: { (error, user) in
                        if let error = error {
                            self.textView.text = error.localizedDescription
                        }else if let user = user {
                            self.textView.text = user.toJSONString(prettyPrint: true)
                            user.firstName = temp
                            XeeRequestManager.shared.updateUser(WithUser: user, completionHandler: nil)
                        }
                    })
                }
            })
            break
        // Get Vehicles
        case 4:
            XeeRequestManager.shared.getVehicles(WithUserID: nil, completionHandler: { (error, vehicules) in
                if let error = error {
                    self.textView.text = error.localizedDescription
                }else if let vehicules = vehicules {
                    let vehicule = vehicules[0]
                    self.vehiculeID = vehicule.vehiculeID!
                    self.deviceID = vehicule.device?.deviceID!
                    self.textView.text = vehicules.toJSONString(prettyPrint: true)
                }
            })
            break
        // Get Vehicule by ID
        case 5:
            if let vehicleID = vehiculeID {
                XeeRequestManager.shared.getVehicle(WithVehicleID: vehicleID, completionHandler: { (error, vehicule) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let vehicule = vehicule {
                        self.textView.text = vehicule.toJSONString(prettyPrint: true)
                    }
                })
            }
            break
        // Get Vehicle Status
        case 6:
            if let vehicleID = vehiculeID {
                XeeRequestManager.shared.getStatus(WithVehicleID: vehicleID, completionHandler: { (error, status) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let status = status {
                        self.textView.text = status.toJSONString(prettyPrint: true)
                    }
                })
            }
            break
        // Update vehicle
        case 7:
            if let vehicleID = vehiculeID {
                XeeRequestManager.shared.getVehicle(WithVehicleID: vehicleID, completionHandler: { (error, vehicle) in
                    if let vehicle = vehicle {
                        let temp = vehicle.name
                        vehicle.name = "TEST"
                        XeeRequestManager.shared.updateVehicle(WithVehicle: vehicle, completionHandler: { (error, vehicle) in
                            if let error = error {
                                self.textView.text = error.localizedDescription
                            }else if let vehicle = vehicle {
                                self.textView.text = vehicle.toJSONString(prettyPrint: true)
                                vehicle.name = temp
                                XeeRequestManager.shared.updateVehicle(WithVehicle: vehicle, completionHandler: nil)
                            }
                        })
                    }
                })
            }
            break
        // Get Vehicle Device
        case 8:
            if let vehicleID = vehiculeID {
                XeeRequestManager.shared.getDevice(ForVehicleID: vehicleID, completionHandler: { (error, device) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let device = device {
                        self.textView.text = device.toJSONString(prettyPrint: true)
                    }
                })
            }
            break
        // Get Vehicle Privacies
        case 9:
            if let vehicleID = vehiculeID {
                XeeRequestManager.shared.getPrivacies(ForVehicleID: vehicleID, From: nil, To: nil, Limit: nil, completionHandler: { (error, privacies) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let privacies = privacies {
                        self.privacyID = privacies[0].privacyID
                        self.textView.text = privacies.toJSONString(prettyPrint: true)
                    }
                })
            }
            break
        // Start Privacy
        case 10:
            if let vehicleID = vehiculeID {
                XeeRequestManager.shared.startPrivacy(ForVehicleID: vehicleID, completionHandler: { (error, privacy) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let privacy = privacy {
                        self.privacyID = privacy.privacyID
                        self.textView.text = privacy.toJSONString(prettyPrint: true)
                    }
                })
            }
            break
        // Stop Privacy
        case 11:
            if let privacyID = privacyID {
                XeeRequestManager.shared.stopPrivacy(ForVPrivacyID: privacyID, completionHandler: { (error, privacy) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let privacy = privacy {
                        self.textView.text = privacy.toJSONString(prettyPrint: true)
                    }
                })
            }else {
                self.textView.text = "No privacy created"
            }
            break
        // Vehicle Trips
        case 12:
            if let vehicleID = vehiculeID {
                XeeRequestManager.shared.getTrips(WithVehicleID: vehicleID, completionHandler: { (error, trips) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let trips = trips {
                        self.tripID = trips[0].tripID
                        self.textView.text = trips.toJSONString(prettyPrint: true)
                    }
                })
            }
            break
        // Vehicle Trip
        case 13:
            if let tripID = tripID {
                XeeRequestManager.shared.getTrip(WithTripID: tripID, completionHandler: { (error, trip) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let trip = trip {
                        self.textView.text = trip.toJSONString(prettyPrint: true)
                    }
                })
            }
            break
        // Trip Signals
        case 14:
            if let tripID = tripID {
                XeeRequestManager.shared.getSignals(WithTripID: tripID, completionHandler: { (error, signals) in
                    if let error = error {
                        self.textView.text = error.localizedDescription
                    }else if let signals = signals {
                        self.textView.text = signals.toJSONString(prettyPrint: true)
                    }
                })
            }
            break
        // Disconnect
        case 15:
            XeeConnectManager.shared.disconnect()
            break
        default:
            break
        }
    }
    
    @IBAction func edit(_ sender: UIButton) {
        let alert: UIAlertController = UIAlertController(title: "Edit", message: "Quelle est la nouvelle valeur ?", preferredStyle: .alert)
        alert.addTextField(configurationHandler: nil)
        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
            if let newID = alert.textFields?[0].text {
                switch sender.tag {
                case 1:
                    self.userID = newID
                    break
                case 2:
                    self.vehiculeID = newID
                    break
                case 3:
                    self.tripID = newID
                    break
                case 4:
                    self.deviceID = newID
                    break
                default:
                    break
                }
            }
        }))
        present(alert, animated: true, completion: nil)
    }
    
    // MARK: - XeeConnectManagerDelegate
    
    func didSuccess(token: XeeToken) {
        self.textView.text = token.toJSONString(prettyPrint: true)
    }
    
    func didFail(WithError error: Error) {
        self.textView.text = error.localizedDescription
    }
    
    func didCancel() {
        self.textView.text = "Canceled"
    }
    
    func didDisconnected() {
        self.textView.text = "Disconnected"
    }

}
